cpp/src/lists/contains.cu

```diff
@@ -105,48 +140,168 @@ struct search_list_fn {
+                      static_cast<rhs_index_type>(search_key_is_scalar ? 0 : list.row_index()));
+      });
+    // If the key is found, return its found position in the list from `found_iter`.
+    return found_iter == end ? NOT_FOUND_SENTINEL : *found_iter;
+  }
+};
+
+/**
+ * @brief Function to search for key element(s) in the corresponding rows of a lists column,
+ * specialized for non-nested types.
+ */
+template <bool search_key_is_scalar,
+          typename Element,
+          typename InputIterator,
+          typename SearchKeyType>
+void index_of_non_nested(InputIterator input_it,
+                         size_type num_rows,
+                         size_type* output_it,
+                         SearchKeyType const& search_keys,
+                         bool search_keys_have_nulls,
+                         duplicate_find_option find_option,
+                         rmm::cuda_stream_view stream)
+{
+  auto const do_search = [&](auto const keys_iter) {
+    thrust::transform(rmm::exec_policy(stream),
+                      input_it,
+                      input_it + num_rows,
+                      keys_iter,
+                      output_it,
+                      search_list_non_nested_fn{find_option});
+  };
+
+  if constexpr (search_key_is_scalar) {
+    auto const keys_iter = cudf::detail::make_optional_iterator<Element>(
+      search_keys, nullate::DYNAMIC{search_keys_have_nulls});
+    do_search(keys_iter);
+  } else {
+    auto const keys_cdv_ptr = column_device_view::create(search_keys, stream);
+    auto const keys_iter    = cudf::detail::make_optional_iterator<Element>(
+      *keys_cdv_ptr, nullate::DYNAMIC{search_keys_have_nulls});
+    do_search(keys_iter);
+  }
+}
+
+/**
+ * @brief Function to search for index of key element(s) in the corresponding rows of a lists
+ * column, specialized for nested types.
+ */
+template <bool search_key_is_scalar, typename InputIterator, typename SearchKeyType>
+void index_of_nested_types(InputIterator input_it,
+                           size_type num_rows,
+                           size_type* output_it,
+                           column_view const& child,
+                           SearchKeyType const& search_keys,
+                           duplicate_find_option find_option,
+                           rmm::cuda_stream_view stream)
+{
+  // Create a `table_view` from the search key(s).
+  // If the input search key is a (nested type) scalar, a new column is materialized from that
+  // scalar before a `table_view` is generated from it. As such, the new created column will also be
+  // returned to keep the result `table_view` valid.
+  [[maybe_unused]] auto const [keys_tview, unused_column] =
+    [&]() -> std::pair<table_view, std::unique_ptr<column>> {
+    if constexpr (std::is_same_v<SearchKeyType, cudf::scalar>) {
+      auto tmp_column       = make_column_from_scalar(search_keys, 1, stream);
+      auto const keys_tview = tmp_column->view();
```

Comment:
Sorry, this part of it looks expensive. Could we not attempt this without having to blow up the scalar into a vector? The non-nested code path does not require this blow-up.

I would have thought this doable with an appropriate iterator.

Reply:
> I would have thought this doable with an appropriate iterator.

Hmm... The row-comparator uses `table_view`, so this can't be done naively. We would need a table/column_view that mimics what the `optional_comparator` does for `scalar` values. It would be non-trivial.

I'm fine with this being merged, in the name of expediency. But I'd like for us to revisit this issue, to avoid having to blow up a scalar into a vector, simply because of an unfortunate interface.

Reply:
@GregoryKimball we talked about this before, when we want to compare a column (table) with a scalar but the existing comparator can only compare tables. Now we have more voice on this, so I think we should prioritize this somehow.

Reply:
I think that comparators for columns/tables against scalars are a reasonable feature request. We can probably find a way to implement this efficiently with a wrapper of some kind around the experimental row operators. Can you file an issue, and list all the places where you know this would be useful?

Reply:
Upon further investigation: the current state is quite yucky, with lifetime management for one code path but not the other. We should be able to handle this a different way, by making a comparator that accepts a scalar, owns the temporary column, and accepts one index (for the column/table input). Centralizing lifetime management in the comparator has been part of our existing designs, so that's probably good to continue.

There aren't a lot of alternatives I can see without improving this at the comparator level, but it's pretty gross in the current state.

Reply:
https://github.com/rapidsai/cudf/issues/10892#issuecomment-1217233559

@ttnghia If you can, please add any other cases to this issue where you know this problem is occurring. There are some alternatives I can imagine to the current behavior, but none of them are as good as solving the problem at the comparator level.

Reply:
> Can you file an issue, and list all the places where you know this would be useful?

Sorry I forgot to mention the already filed issue https://github.com/rapidsai/cudf/issues/10892. Thanks for bringing it up.